libtenzir/builtins/functions/string.cpp

```diff
@@ -197,6 +197,135 @@ class nullary_method : public virtual method_plugin {
+    auto replacement = std::string{};
+    auto max_replacements = std::optional<located<int64_t>>{};
+    TRY(argument_parser2::method(name())
+          .add(subject_expr, "<string>")
+          .add(pattern, "<pattern>")
+          .add(replacement, "<replacement>")
+          .add("max", max_replacements)
+          .parse(inv, ctx));
+    if (max_replacements) {
+      if (max_replacements->inner < 0) {
+        diagnostic::error("`max` must be at least 0, but got {}",
+                          max_replacements->inner)
+          .primary(*max_replacements)
+          .emit(ctx);
+      }
+    }
+    return function_use::make(
+      [this, subject_expr = std::move(subject_expr),
+       pattern = std::move(pattern), replacement = std::move(replacement),
+       max_replacements](evaluator eval, session ctx) -> series {
+        auto result_type = string_type{};
+        auto result_arrow_type
+          = std::shared_ptr<arrow::DataType>{result_type.to_arrow_type()};
+        auto subject = eval(subject_expr);
+        auto f = detail::overload{
+          [&](const arrow::StringArray& array) {
+            auto max = max_replacements ? max_replacements->inner : -1;
+            auto options = arrow::compute::ReplaceSubstringOptions(
+              pattern, replacement, max);
+            auto result = arrow::compute::CallFunction("replace_substring",
+                                                       {array}, &options);
+            if (not result.ok()) {
+              diagnostic::warning("{}", result.status().ToString())
+                .primary(subject_expr)
+                .emit(ctx);
+              return series::null(result_type, subject.length());
+            }
+            return series{result_type, result.MoveValueUnsafe().make_array()};
+          },
+          [&](const arrow::NullArray& array) {
+            return series::null(result_type, array.length());
+          },
+          [&](const auto&) {
+            diagnostic::warning("`{}` expected `string`, but got `{}`", name(),
+                                subject.type.kind())
+              .primary(subject_expr)
+              .emit(ctx);
+            return series::null(result_type, subject.length());
+          },
+        };
+        return caf::visit(f, *subject.array);
+      });
+  }
+};
+
+class slice : public virtual method_plugin {
+public:
+  auto name() const -> std::string override {
+    return "tql2.slice";
+  }
+
+  auto make_function(invocation inv, session ctx) const
+    -> failure_or<function_ptr> override {
+    auto subject_expr = ast::expression{};
+    auto begin = std::optional<located<int64_t>>{};
+    auto end = std::optional<located<int64_t>>{};
+    auto stride = std::optional<located<int64_t>>{};
+    TRY(argument_parser2::method(name())
+          .add(subject_expr, "<string>")
+          .add("begin", begin)
+          .add("end", end)
+          .add("stride", stride)
+          .parse(inv, ctx));
```

Comment:
@mavam I think @dominiklohmann has made a very good point with making everything optional. I do like being able to specify only the end. However, what I don't like is that I cannot call `x = y.slice(7, 13)` anymore. I must type `x = y.slice(begin=3, end=7)` to make it work. 

|             | Positional   | Named                  |
|-------------|--------------|------------------------|
| Start       | `.slice(3)`    | `.slice(start=3)`        |
| Start & End | `.slice(3, 7)` | `.slice(start=3, end=7)` |
| End         | `.slice(0, 7)` | `.slice(end=7)`          |

We could put some work into the argument parser to support parameters that support both styles. But I'm not sure about that. Makes me almost want to add some native syntax like `x[3:7]` or `x[3..7]`... 

Thoughts?

Reply:
Great points. I factored the discussion and your proposal into https://github.com/tenzir/issues/issues/2138, as this one is working right now and also behaving like the existing `slice` operator.